### PR TITLE
BTB-46-product-page

### PR DESCRIPTION
**Pull request by heiligerhuegel**
# Issue
https://amiconsult.atlassian.net/browse/BTB-46

## Changes

- Added Single Product Page 
- added tests for single Product Page

## Checklist

* [x] I tested the functionality on my local environment.
* [x] I made sure my code is easy to read and understand.
* [x] I provided sufficient automated tests for the code.
* [x] There a no conflicts from my-branch..main.

### CODE REVIEW

File: src/app/api/oneproduct.service.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  oneProduct$: Observable<IOneSingleProduct | undefined> =
+    this.oneProductSubject.asObservable();
+
+  setOneProductSubject(newValue: IOneSingleProduct | undefined) {
+    this.oneProductSubject.next(newValue);
+  }
+
+  constructor(private http: HttpClient, private country: CountryService) {}
+
+  getSingleProductDetails(
```
Comment (jolle22) [inline src/app/api/oneproduct.service.ts:28]: Would it be okay for you to rename this to getProductDetails?
The fact that its a only one product is stated by "Product" which is singular. If the method would return more then one product the name should contain the plural "Products".

File: src/app/Interface/singleproduct.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
+export interface IOneSingleProduct {
```
Comment (jolle22) [inline src/app/Interface/singleproduct.ts:1]: Would it be okay for you to rename this to IProduct?
The fact that its a only one product is stated by "Product" which is singular. If the method would return more then one product the name should contain the plural "Products".

**Review by jolle22 — CHANGES_REQUESTED** [review]

File: src/app/components/product-detail/product-detail.component.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+  ngOnInit(): void {
+    this.oneProductService
+      .getSingleProductDetails(this.route.snapshot.paramMap.get('id') || '')
+      .pipe(takeUntil(this.destroy$))
+      .subscribe({
+        next: (data) => {
+          this.oneProduct = data;
+        },
+      });
```
Comment (jolle22) [inline src/app/components/product-detail/product-detail.component.ts:53]: Please move this out to a separate function and call it from ngOnInit.

Like this you get the chance to use a descriptive function name. If the name says what the function does I don't need to read the code to understand what is happening.
Also testing the code becomes easier.

File: src/app/api/oneproduct.service.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
+import { HttpClient, HttpErrorResponse } from '@angular/common/http';
+import { Injectable } from '@angular/core';
+import { BehaviorSubject, Observable, catchError, throwError } from 'rxjs';
+import { IOneSingleProduct } from '../Interface/singleproduct';
+import { CountryService } from '../services/country.service';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class OneProductService {
```
Comment (jolle22) [inline src/app/api/oneproduct.service.ts:10]: Would it be okay for you to rename this to ProductService?
The fact that its a only one product is stated by "Product" which is singular. If the method would return more then one product the name should contain the plural "Products".

File: src/app/api/oneproduct.service.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+  oneProduct!: IOneSingleProduct | undefined;
+
+  private oneProductUrl =
```
Comment (jolle22) [inline src/app/api/oneproduct.service.ts:13]: Would it be okay for you to rename this to productUrl?

File: src/app/api/oneproduct.service.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    'https://ac-trainee-store-api.herokuapp.com/products/';
+
+  private oneProductSubject: BehaviorSubject<IOneSingleProduct | undefined> =
```
Comment (jolle22) [inline src/app/api/oneproduct.service.ts:16]: Would it be okay for you to rename this to productSubject?

**Review by jolle22 — CHANGES_REQUESTED** [review]

File: src/app/Interface/singleproduct.ts (hunk 2 above)
Comment (heiligerhuegel) [reply src/app/Interface/singleproduct.ts:1]: the name "IProduct" is already taken, by another interface, the Interface was renamed to "IDetailedProduct" and now it extends the "IProduct" Interface.

File: src/app/api/oneproduct.service.ts (hunk 1 above)
Comment (heiligerhuegel) [reply src/app/api/oneproduct.service.ts:28]: "getSingleProductDetails"  was renamed to "getProductDetails"

File: src/app/components/product-detail/product-detail.component.ts (hunk 3 above)
Comment (heiligerhuegel) [reply src/app/components/product-detail/product-detail.component.ts:53]: created a new function called "getProductDetailsById" which is now called in the ngOnInit function.

File: src/app/api/oneproduct.service.ts (hunk 5 above)
Comment (heiligerhuegel) [reply src/app/api/oneproduct.service.ts:13]: oneProductUrl was renamed to productUrl

File: src/app/api/oneproduct.service.ts (hunk 4 above)
Comment (heiligerhuegel) [reply src/app/api/oneproduct.service.ts:10]: OneProductService was renamed to ProductService

File: src/app/api/oneproduct.service.ts (hunk 6 above)
Comment (heiligerhuegel) [reply src/app/api/oneproduct.service.ts:16]: oneProductSubject was renamed to productSubject

**Review by jolle22 — APPROVED** [review]

**Pull request merged**